# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [4]:
def addtwo(x):
    return x+2
        


# as lambda 

lambda x: x+2

print((lambda x: x+2)(3))

5


#### Saving to a Variable

In [7]:
f = lambda x, y: x + y + 2
print(f(2,4))

8


#### Multiple Inputs

In [8]:
(lambda x,y,z:x*y*z)(3,5,8)

120

#### Passing a Lambda into a Function

In [11]:
def math(func,num):
    return func(num)

math(lambda x:x*x, 5)

25

#### Returning a Lambda from a Function

In [13]:
def returnfunc():
    def multi(num):
        return num*num
    return multi
returnfunc()(4)

16

In [18]:
def returnlamb(b,c):
    return lambda x,y:(x+y)-(b+c)

returnlamb(4,3)(3,2)

-2

#### If Statements within Lambdas

In [20]:
#Lambda x : True if (condition) else false

f = lambda num: num*2 if num > 10 else num + 2 
f(12)
f(10)

12

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [21]:
f = lambda n: n**3
f(3)

27

In [22]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [23]:
sorted([344,545,7,45544,64,34,23,43,3535,54])

[7, 23, 34, 43, 54, 64, 344, 545, 3535, 45544]

In [30]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest
w = {
    "name": "Wade",
    "age": 223
}
z={
    "name":"Zak",
    "age":14
}
c = {
    "name":"Calen",
    "age":3
}
k={
    "name":"Kristen",
    "age":21
}
people = [w,k,z,c]
print(people)
myl=lambda person_dict: person_dict["age"]
print(sorted(people,key=myl))

print(sorted(people, key=lambda person_dict: person_dict["name"],reverse=True))

[{'name': 'Wade', 'age': 223}, {'name': 'Kristen', 'age': 21}, {'name': 'Zak', 'age': 14}, {'name': 'Calen', 'age': 3}]
[{'name': 'Calen', 'age': 3}, {'name': 'Zak', 'age': 14}, {'name': 'Kristen', 'age': 21}, {'name': 'Wade', 'age': 223}]
[{'name': 'Zak', 'age': 14}, {'name': 'Wade', 'age': 223}, {'name': 'Kristen', 'age': 21}, {'name': 'Calen', 'age': 3}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [31]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [32]:
def squared(num,num2):
    if num < 10 and num2 < 10:
        return num**2,num2**2
    else:
        return num, num2
numbers = [4,29,3,4,2]
morenum = [3,11,20,3,15]
squarenummap = map(squared, numbers , morenum)
print(list(squarenummap))

[(16, 9), (29, 11), (3, 20), (16, 9), (2, 15)]


#### Using Lambda's with Map

In [34]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
numbers= [4,10,3,2,6]
squarenums = list(map((lambda x: x**2 if x<10 else x),numbers))
print(squarenums)

[16, 10, 9, 4, 36]


In [39]:
squarenum2 = list(map((lambda x1, y1: (x1**2, y1**2) if x1<10 and y1<10 else (x1,y1)),numbers,morenum))
squarenum2

[(16, 9), (10, 11), (3, 20), (4, 9), (6, 15)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [42]:
numbers= [4,8,2,6]
doubleminus = list(map((lambda x: x*2 , x-1),numbers))
doubleminus

NameError: name 'x' is not defined

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [43]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Using Lambda's with Filter()

In [45]:
names = ["Krist","Alex","Kaleb","Sticker","Kevin","David"]

def knames(name):
    if name[0].lower() == "k":
        return True
    else:
        return False
    
newlist=list(filter(knames,names))
print(newlist)

['Krist', 'Kaleb', 'Kevin']


In [49]:
new_namelamb = list(filter(lambda name:True if name[0].lower() == "k" else False, names))
print(new_namelamb)

['Krist', 'Kaleb', 'Kevin']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [58]:
from statistics import mean
mean([2,4,6,4,33])


9.8

In [59]:
nums = [2,5,88,88,756,3,54,33]
belowmean = list(filter(lambda x: True if x >= mean(nums) else False ,nums))
print (mean(nums))
print(belowmean)

128.625
[756]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [61]:
from functools import reduce
# reduce(function,iterable)

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



#### Using Lambda's with Reduce()

In [62]:
list1 = [2,4,6,8,10]
def subnum(num1,num2):
    return num1 - num2
result = reduce(subnum, list1)
result


-26

In [64]:
l3 = [1,10,2,30]
reslamb = reduce(lambda num1, num2:num1-num2 ,l3)
reslamb

-41

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [65]:
l5 = [1,2,3,4]
multilamb = reduce(lambda x,y:x*y,l5)
multilamb

24

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [2]:
def recursive(i):
    print("helo")
    if i == 4:
        return
    else:
        i+=1
        recursive(i)
recursive(0)

helo
helo
helo
helo
helo


#### Implementing a Base Case

In [6]:
def addnums(num):
    #set base case (point when we want to stop)
    if num <= 1:
        print("addnums(1)")
        return num
    else:
        print(f"addnums({num})")
        x= num-1
        return num + addnums(x)
addnums(5)

addnums(5)
addnums(4)
addnums(3)
addnums(2)
addnums(1)


15

#### Writing a Factorial Function

In [9]:
# 5! = 5*4*3*2*1
def factorial(num):
    if num <= 1:
        print('factorial(1) = 1')
        return 1
    else:
        print(f"factorial({num}) = {num} * factorial({num-1})")
        return num * factorial(num-1)
factorial(5)

factorial(5) = 5 + factorial(4)
factorial(4) = 4 + factorial(3)
factorial(3) = 3 + factorial(2)
factorial(2) = 2 + factorial(1)
factorial(1) = 1


120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [13]:
n = 5

def subnums(num):
    if num <=1:
        return 1
    else:
        return num - subnums(num-1)
    
subnums(5)

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [15]:
list(range(6))

[0, 1, 2, 3, 4, 5]

In [16]:
def test():
    yield 1
    yield 2
    yield 3
    yield 'a'
    yield ['a','b','c']
    
for i in test():
    print(i)

1
2
3
a
['a', 'b', 'c']


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [18]:
def myrange(n):
    for i in range(n):
        yield i
list(myrange(5))

[0, 1, 2, 3, 4]

In [19]:
def myrange(start,stop,step=1):
    while start<stop:
        yield start
        start+=step
list(myrange(1,5,1))

[1, 2, 3, 4]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [23]:
def therange(start,stop,step=1):
    while start > stop:
        yield start **2
        start -= step
        
for num in therange(0,20,1):
    print(num)

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [27]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

In [35]:
new_places = list(filter(lambda places:True if places.strip() != "" else False, places))
print(new_places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [36]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


In [3]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
sorted_author = sorted(author, key=lambda s: s.upper().split()[-1])
print(sorted_author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [38]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


<class 'str'>


In [2]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

cel_to_far = list(map(lambda x: (x[0], (9/5) * x[1] + 32), places))
print(cel_to_far)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [1]:
def fib(z,num1=0,num2=1,x=0):
    if x >= z :
        print(f'Iteration {x}: {num2}')
        return None
    else:
        print(f'Iteration {x}: {num2}')
        x += 1
        y = num2
        num2 = num1 + num2
        num1 = y
        
        return fib(z,num1,num2,x)

fib(5)

Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
